<img src="https://github.com/insaid2018/Term-1/blob/master/Images/INSAID_Full%20Logo.png?raw=true" width="240" height="360" />

# Insurance Dataset_Term3_Project1 (Pycaret)

### Problem Statement : - Automating ML and Finding the best model for the insurance dataset with Pycaret

**We have an altered dataset from Insurance_Dataset_Term3_Project1_Part1 notebook; which has a feature columns as Product_Info, Employment_Info, Family_Hist, InsuredInfo, Medical_History, Medical_Keyword(dummy variables) and Response. And we need to prepare data and deploy efficient model for target variabe (Response).** 

<center><img src="https://pycaret.org/wp-content/uploads/2020/03/Divi93_43.png" width="300" height="80" /></center>

### Features :-
- An Open Source and low-code machine learning Library for data preaparation and deployment of model.
- Time efficient (Less time spent on code and more productive).
- Simple & Easy to use in end-to-end machine learning experaiments with less lines of code.
- It's business ready solution allows to do prototyping quickly and efficiently from notebook environment.

In [4]:
import pycaret                            # import required library
import numpy as np
import pandas as pd

In [2]:
pycaret. __version__

'2.0.0'

***Required dataset:- Insurance_df3***

In [5]:
Ins_df = pd.read_csv('Insurance_df3.csv', index_col=0)         # Insurance dataset (Insurance_df3.csv) for Pycaret model evaluation.
Ins_df.head()

Id  Product_Info_1  Product_Info_3  Product_Info_5  Product_Info_6  \
0   2               1              10               2               1   
1   5               1              26               2               3   
2   6               1              26               2               3   
3   7               1              10               2               3   
4   8               1              26               2               3   

   Product_Info_7  Employment_Info_2  Employment_Info_3  Employment_Info_5  \
0               1                 12                  1                  3   
1               1                  1                  3                  2   
2               1                  9                  1                  2   
3               1                  9                  1                  3   
4               1                  9                  1                  2   

   InsuredInfo_1  ...  Medical_Keyword_41  Medical_Keyword_42  \
0              1  ...                   0                   0   
1              1  ...                   0                   0   
2              1  ...                   0                   0   
3              2  ...                   0                   0   
4              1  ...                   0                   0   

   Medical_Keyword_43  Medical_Keyword_44  Medical_Keyword_45  \
0                   0                   0                   0   
1                   0                   0                   0   
2                   0                   0                   0   
3                   0                   0                   0   
4                   0                   0                   0   

   Medical_Keyword_46  Medical_Keyword_47  Medical_Keyword_48  Response  \
0                   0                   0                   0         8   
1                   0                   0                   0         4   
2                   0                   0                   0         8   
3                   0                   0                   0         8   
4                   0                   0                   0         8   

   Product_Info_2  
0              16  
1               0  
2              18  
3              17  
4              15  

[5 rows x 115 columns]

|Records|Features|Dataset Size|
|:---- | :-----|:-----|
|59381|115|15.2 MB|

|Variable|Description|
| :--  |  :---  |
|Id |A unique identifier associated with an application.|
|Product_Info_1-7 | A set of normalized variables relating to the product applied for |
|Ins_Age | Normalized age of applicant |
|Ht | Normalized height of applicant |
|Wt | Normalized weight of applicant |
|BMI | Normalized BMI of applicant |
|Employment_Info_1-6 | A set of normalized variables relating to the employment history of the applicant. |
|InsuredInfo_1-6 | A set of normalized variables providing information about the applicant. |
|Insurance_History_1-9 | A set of normalized variables relating to the insurance history of the applicant. |
|Family_Hist_1-5 | A set of normalized variables relating to the family history of the applicant. |
|Medical_History_1-41 | A set of normalized variables relating to the medical history of the applicant. |
|Medical_Keyword_1-48 | A set of dummy variables relating to the presence of/absence of a medical keyword being associated with the application.|
|**Response** | **This is the target variable**, an ordinal variable relating to the final decision associated with an application |

***The following variables are all categorical (nominal):***

Product_Info_1, Product_Info_2, Product_Info_3, Product_Info_5, Product_Info_6, Product_Info_7, Employment_Info_2, Employment_Info_3, Employment_Info_5, InsuredInfo_1, InsuredInfo_2, InsuredInfo_3, InsuredInfo_4, InsuredInfo_5, InsuredInfo_6, InsuredInfo_7, Insurance_History_1, Insurance_History_2, Insurance_History_3, Insurance_History_4, Insurance_History_7, Insurance_History_8, Insurance_History_9, Family_Hist_1, Medical_History_2, Medical_History_3, Medical_History_4, Medical_History_5, Medical_History_6, Medical_History_7, Medical_History_8, Medical_History_9, Medical_History_11, Medical_History_12, Medical_History_13, Medical_History_14, Medical_History_16, Medical_History_17, Medical_History_18, Medical_History_19, Medical_History_20, Medical_History_21, Medical_History_22, Medical_History_23, Medical_History_25, Medical_History_26, Medical_History_27, Medical_History_28, Medical_History_29, Medical_History_30, Medical_History_31, Medical_History_33, Medical_History_34, Medical_History_35, Medical_History_36, Medical_History_37, Medical_History_38, Medical_History_39, Medical_History_40, Medical_History_41


***The following variables are continuous:***

Product_Info_4, Ins_Age, Ht, Wt, BMI, Employment_Info_1, Employment_Info_4, Employment_Info_6, Insurance_History_5, Family_Hist_2, Family_Hist_3, Family_Hist_4, Family_Hist_5


***The following variables are discrete:***

Medical_History_1, Medical_History_10, Medical_History_15, Medical_History_24, Medical_History_32

***dummy variables***

Medical_Keyword_1-48 are dummy variables.

### Importing the module

In [4]:
from pycaret.classification import *               # import classification 

### Initializing the setup

In [5]:
clf1 = setup(data = Ins_df,                       # Let's do the setup for Ins_df dataset.
             target = 'Response',
             numeric_imputation = 'zero',
             ignore_features = ['Id'])

Setup Succesfully Completed!


***We can see here that all the setup values and descriptions. After observation we can now compare various classification models.***

In [7]:
compare_models()         # we can use exclude=['catboost'] to save processing time       

GradientBoostingClassifier(ccp_alpha=0.0, criterion='friedman_mse', init=None,
                           learning_rate=0.1, loss='deviance', max_depth=3,
                           max_features=None, max_leaf_nodes=None,
                           min_impurity_decrease=0.0, min_impurity_split=None,
                           min_samples_leaf=1, min_samples_split=2,
                           min_weight_fraction_leaf=0.0, n_estimators=100,
                           n_iter_no_change=None, presort='deprecated',
                           random_state=7028, subsample=1.0, tol=0.0001,
                           validation_fraction=0.1, verbose=0,
                           warm_start=False)

**It can be seen from model camparison above that, Gradient Boosting Classifier in accuracy score, precision score, f1 score. And hence we will create model for Gradient Boosting Classifier***

## Model 1: Gradient Boosting Classifier (gbc)

In [6]:
gbc  = create_model('gbc')      # Let's create model for gbc (short name for Gradient boosting classifier)

**Here we can compare model creation of 10 folds(default), mean and standard deviation value for gradient boosting classifier (gbc) model.**

In [7]:
gbc         # gbc model

GradientBoostingClassifier(ccp_alpha=0.0, criterion='friedman_mse', init=None,
                           learning_rate=0.1, loss='deviance', max_depth=3,
                           max_features=None, max_leaf_nodes=None,
                           min_impurity_decrease=0.0, min_impurity_split=None,
                           min_samples_leaf=1, min_samples_split=2,
                           min_weight_fraction_leaf=0.0, n_estimators=100,
                           n_iter_no_change=None, presort='deprecated',
                           random_state=5370, subsample=1.0, tol=0.0001,
                           validation_fraction=0.1, verbose=0,
                           warm_start=False)

**Let's tuned the model for best results of various score.**

In [8]:
tuned_gbc = tune_model(gbc)      # tune gbc model 

**Further step in prediction we can apply predict_model on tuned_gbc model**

In [9]:
predictions_clfr = predict_model(tuned_gbc)  # predictions of model

Model  Accuracy  AUC  Recall   Prec.      F1  \
0  Gradient Boosting Classifier     0.412    0  0.2463  0.3718  0.3772   

    Kappa     MCC  
0  0.2354  0.2401

**In model evaluation technques we can get information on hyper-parameters, ROC curve, Confusion matrix, Precision - Recall, Learning and validation of model, Class report, Feature importance, Decision boundary. and with this information it's possible to evaluate peerformance of model.**

In [10]:
evaluate_model(tuned_gbc)       # Let's see this model_evaluation

interactive(children=(ToggleButtons(description='Plot Type:', icons=('',), options=(('Hyperparameters', 'param…




**After evaluation of Gradient Boosting Classifier (gbc) we can now consider to compare Random forest classfier (rf), Decision tree classifier (dt) and Logistic regression (lr).**

## Let's compare 3 models - RF, DT and LR

In [6]:
compare_models(include=['lr','dt','rf'])  

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=10, n_jobs=-1,
                       oob_score=False, random_state=6070, verbose=0,
                       warm_start=False)

**Observation: - with the above observation it can be seen that RF model is better in comparision of accuracy, Precision-Recall and F1 Score. So, Random Forest classfier (rf) is our first model in prepartion.**

## Model 2 :- Random Forest Classifier (rf)

In [7]:
rf = create_model('rf')     # create model

In [8]:
rf

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=10, n_jobs=-1,
                       oob_score=False, random_state=6070, verbose=0,
                       warm_start=False)

**Let's tuned the random forest classifier model.**

In [9]:
tuned_rf = tune_model(rf)

In [10]:
predictions_clsrf = predict_model(tuned_rf)       # Prediction of RF model

Model  Accuracy  AUC  Recall   Prec.      F1   Kappa  \
0  Random Forest Classifier    0.4533    0  0.2708  0.4288  0.4005  0.2743   

      MCC  
0  0.2873

**Random Forest Classifier (rf) Model evaluation with techniques.**

In [11]:
evaluate_model(tuned_rf)

interactive(children=(ToggleButtons(description='Plot Type:', icons=('',), options=(('Hyperparameters', 'param…

**Random Forest Classifier Observation:-**

- **Confusion Matrix :-** We can compare True class & Predicted class for 1-8 response given.
- **Precision-Recall :-** We can see precision to recall score for model evaluation.
- **Class Report :-** We can clssification report of response given with precision, recall, f1 score and support.
- **Feature Importance :-** We can compare feature importance like Medical_History, InsuredInfo, Medical_Keyword and so on.
- **ROC-AUC curve :-** We can compare ROC-AUC curve for 1-8 response.
- **Learning curve :-** We can compare Training and Cross validation score with Training instances.
- **Validation curve :-** We can compare Training and Cross validation score with Max_depth.


## Model 3 :- Decision Tree Classifier (DT)

In [12]:
dt = create_model('dt')           # Create modeling

In [13]:
dt

DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='gini',
                       max_depth=None, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=6070, splitter='best')

In [14]:
tuned_dt = tune_model(dt)          # Let's tuned the dt model

In [15]:
predictions_clsdt = predict_model(tuned_dt)            # Prediction of dt model

Model  Accuracy  AUC  Recall   Prec.     F1   Kappa  \
0  Decision Tree Classifier    0.4157    0  0.2437  0.3539  0.357  0.2222   

      MCC  
0  0.2351

**Decision Tree Classifier (dt) Model evaluation with techniques.**


In [26]:
evaluate_model(tuned_dt)

interactive(children=(ToggleButtons(description='Plot Type:', icons=('',), options=(('Hyperparameters', 'param…

**Decision Tree Classifier Observation:-**

- **Confusion Matrix :-** We can compare True class & Predicted class for 1-8 response given.
- **Precision-Recall :-** We can see precision to recall score for model evaluation.
- **Class Report :-** We can clssification report of response given with precision, recall, f1 score and support.
- **Feature Importance :-** We can compare feature importance like Medical_History, InsuredInfo, Family_Hist, Product_Info, Medical_Keyword and so on.
- **ROC-AUC curve :-** We can compare ROC-AUC curve for 1-8 response.
- **Learning curve :-** We can compare Training and Cross validation score with Training instances.
- **Validation curve :-** We can compare Training and Cross validation score with Max_depth.

## Model 4:- Logistic Regression (lr)

In [31]:
lr = create_model('lr')           # create model

In [18]:
lr

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=6070, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

In [21]:
tuned_lr = tune_model(lr)           # Let's tune the lr model

In [22]:
predictions_clslr = predict_model(tuned_lr)          # Prediction of lr model

Model  Accuracy  AUC  Recall   Prec.      F1   Kappa     MCC
0  Logistic Regression    0.3671    0  0.1726  0.2953  0.2495  0.1079  0.1445

**Logistic Regression (lr) Model evaluation with techniques.**

In [23]:
evaluate_model(tuned_lr)

interactive(children=(ToggleButtons(description='Plot Type:', icons=('',), options=(('Hyperparameters', 'param…

**Logistic Regression Observation:-**

- **Confusion Matrix :-** We can compare True class & Predicted class for 1-8 response given.
- **Precision-Recall :-** We can see precision to recall score for model evaluation.
- **Class Report :-** We can clssification report of response given with precision, recall, f1 score and support.
- **Feature Importance :-** We can compare feature importance like Employment_Info, Product_Info, Medical_History, Medical_Keyword and so on.
- **ROC-AUC curve :-** We can compare ROC-AUC curve for 1-8 response.
- **Learning curve :-** We can compare Training and Cross validation score with Training instances.
- **Validation curve :-** We can compare Training and Cross validation score with Max_depth.

# Conclusion :-

**With the help of Pycaret Model Evaluation we can easily compare classification models and also with Hyper-parameter tuning and comparability of score, model selection, preparation and evaluation best possible outcome is possible in quick manner.**

# Thanks!!!